File: nips/tf_ops/build_assignment/build_assignment.cc
#include <string.h>
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/framework/op_kernel.h"

using namespace tensorflow;

/*
Build an assignment cost matrix from image pixel probabilities.

x: BxHxWxN, 4-D probability map of an image for per-pixel class prediction.
y: BxHxW, 3-D label for each pixel.
cost_matrix: BxNxN, 3-D cost matrix for assignment matrix,
  cost_matrix[:,i,j] is the score that pred-class-i matches pred-class-j
*/
REGISTER_OP("BuildAssignment")
    .Input("x: float32")
    .Input("y: int32")
    .Output("cost_matrix: float32");
REGISTER_OP("BuildAssignmentGrad")
    .Input("x: float32")
    .Input("y: int32")
    .Output("grad: float32"); // grad is for x

void build_assignment_matrix(int b, int h, int w, int n, const float* x, const int* y, float* cost_matrix) {
	for (int k=0;k<b;k++){
		for (int i=0;i<h;i++){
			for (int j=0;j<w;j++){
				for (int l=0;l<n;l++){
					cost_matrix[l*n+y[i*w+j]] += x[i*(w*n)+j*n+l];
				}
			}
		}
		x += h*w*n;
		y += h*w;
		cost_matrix += n*n;
	}
}

//void build_assignment_matrix_grad(int b, int h, int w, int n, const float* x, const int* y, float* grad) {
//	for (int k=0;k<b;k++){
//		for (int i=0;i<h;i++){
//			for (int j=0;j<w;j++){
//				for (int l=0;l<n;l++){
//					//grad[i*(w*n)+j*n+l]= cost_matrix_grad[l*n+y[i*w+j]];
//				}
//			}
//		}
//		x += h*w*n;
//		y += h*w;
//		grad += h*w*n;
//	}
//}

class BuildAssignmentOp : public OpKernel {
	public:
		explicit BuildAssignmentOp(OpKernelConstruction* context) : OpKernel(context) {}

		void Compute(OpKernelContext* context) override {
			const Tensor& x_tensor = context->input(0);
			OP_REQUIRES(context, x_tensor.dims() == 4, errors::InvalidArgument("input must be 4-D BxHxWxN."));
			auto x_flat = x_tensor.flat<float>();
			const float* x = &(x_flat(0));
			int b = x_tensor.shape().dim_size(0);
			int h = x_tensor.shape().dim_size(1);
			int w = x_tensor.shape().dim_size(2);
			int n = x_tensor.shape().dim_size(3);

			const Tensor& y_tensor = context->input(1);
			OP_REQUIRES(context, y_tensor.dims() == 3, errors::InvalidArgument("input must be 3-D BxHxW."));
			auto y_flat = y_tensor.flat<int>();
			const int* y = &(y_flat(0));

			Tensor* cost_matrix_tensor = NULL;
			OP_REQUIRES_OK(context, context->allocate_output(0, TensorShape{b,n,n}, &cost_matrix_tensor)); 
			auto cost_matrix_flat = cost_matrix_tensor->flat<float>();
			float* cost_matrix = &(cost_matrix_flat(0));
         		memset(cost_matrix, 0, sizeof(float)*b*n*n);

			build_assignment_matrix(b,h,w,n,x,y,cost_matrix);
		}
};

REGISTER_KERNEL_BUILDER(Name("BuildAssignment").Device(DEVICE_CPU), BuildAssignmentOp);


//class BuildAssignmentGradOp: public OpKernel {
//  public:
//    explicit BuildAssignmentGradOp(OpKernelConstruction* context) : OpKernel(context) {}
//    void Compute(OpKernelContext* context) override {
//	const Tensor& x_tensor = context->input(0);
//	OP_REQUIRES(context, x_tensor.dims() == 4, errors::InvalidArgument("input must be 4-D BxHxWxN."));
//      	auto x_flat = x_tensor.flat<float>();
//      	const float* x = &(x_flat(0));
//      	int b = x_tensor.shape().dim_size(0);
//      	int h = x_tensor.shape().dim_size(1);
//      	int w = x_tensor.shape().dim_size(2);
//      	int n = x_tensor.shape().dim_size(3);
//
//      	const Tensor& y_tensor = context->input(1);
//      	OP_REQUIRES(context, y_tensor.dims() == 3, errors::InvalidArgument("input must be 3-D BxHxW."));
//      	auto y_flat = y_tensor.flat<int>();
//      	const int* y = &(y_flat(0));
//
//	Tensor* grad_tensor = NULL;
//	OP_REQUIRES_OK(context, context->allocate_output(0, TensorShape{b,h,w,n}, &grad_tensor));
//	auto grad_flat = grad_tensor->flat<float>();
//	float* grad = &(grad_flat(0));
//        build_assignment_matrix_grad(b,h,w,n,x,y,grad);
//    }
//};
//
//REGISTER_KERNEL_BUILDER(Name("BuildAssignmentGrad").Device(DEVICE_CPU), BuildAssignmentGradOp);

File: nips/tf_ops/build_assignment/build_assignment.py
import numpy as np
import tensorflow as tf
from tensorflow.python.framework import ops
import os
import sys
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append(BASE_DIR)
build_assignment_module = tf.load_op_library(os.path.join(BASE_DIR, 'build_assignment_so.so'))

def build_assignment(x, y):
  """
  Args
    x: BxHxWxN float tensor of image pixel log probabilities
    y: BxHxW int tensor of image pixel class labels

  Returns
    cost_matrix: BxNxN float tensor
  """
  return build_assignment_module.build_assignment(x, y)
@tf.RegisterShape('BuildAssignment')
def _build_assignment_shape(op):
  shape1 = op.inputs[0].get_shape().with_rank(4)
  shape2 = op.inputs[1].get_shape().with_rank(3)
  return [tf.TensorShape([shape1.dims[0], shape1.dims[3], shape1.dims[3]])]

def my_func(x, y, grad_cost_matrix): # grad_cost_matrix is the grad of the output
  b,h,w,n = x.shape
  grad = np.zeros_like(x).astype('float32')
  for bidx in range(b):
    for i in range(h):
      for j in range(w):
        for k in range(n):
          grad[bidx,i,j,k] = grad_cost_matrix[bidx, k, y[bidx,i,j]]
  return grad

@tf.RegisterGradient('BuildAssignment')
def _build_assignment_grad(op, grad_cost_matrix): # grad is BxNxN
  x = op.inputs[0] # BxHxWxN
  y = op.inputs[1] # BxHxW
  # {grad_x}_i_j_k) = {grad_cost_matrix}_k_yij 
  return [tf.py_func(my_func, [x,y,grad_cost_matrix], tf.float32), None]


if __name__=='__main__':
  with tf.Graph().as_default():
    with tf.device('/cpu:'+str(0)):
      #x = tf.ones((3,5,5,2))
      #y = tf.zeros((3,5,5), dtype=tf.int32)
      x = tf.Variable(np.random.randn(16,2,2,3).astype('float32'))
      y = tf.Variable(np.random.randint(low=0, high=3, size=(16,2,2)).astype('int32'), dtype=tf.int32)
      cost_matrix = build_assignment(x, y)
      print cost_matrix
    with tf.Session('') as sess:
      sess.run(tf.initialize_all_variables())
      x_val, y_val, cost_matrix_val = sess.run([x, y, cost_matrix])
      print x_val[0,...].squeeze()
      print y_val[0,...].squeeze()
      print cost_matrix_val[0,...].squeeze()
      """ Prints out (which is verified as correct):
[[[ 0.47740752  0.29287109 -0.76430809]
  [-0.43345165 -0.17250288 -0.55713695]]

 [[-0.55650556 -0.14719154 -0.28433409]
  [-0.36214039 -0.14317064  0.18925352]]]

[[2 0]
 [2 1]]

[[-0.43345165 -0.36214039 -0.07909805]
 [-0.17250288 -0.14317064  0.14567955]
 [-0.55713695  0.18925352 -1.04864216]]
      """
